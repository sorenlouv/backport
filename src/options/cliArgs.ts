import yargs from 'yargs';
import { OptionsFromConfigFiles } from './config/config';

export type OptionsFromCliArgs = ReturnType<typeof getOptionsFromCliArgs>;
export function getOptionsFromCliArgs(
  configOptions: OptionsFromConfigFiles,
  argv: readonly string[]
) {
  const cliArgs = yargs(argv)
    .parserConfiguration({
      'strip-dashed': true,
      'strip-aliased': true,
      'boolean-negation': false,
    })
    .usage('$0 [args]')
    .wrap(Math.max(100, Math.min(120, yargs.terminalWidth())))
    .option('accessToken', {
      alias: 'accesstoken',
      description: 'Github access token',
      type: 'string',
    })
    .option('all', {
      default: configOptions.all,
      description: 'List all commits',
      alias: 'a',
      type: 'boolean',
    })
    .option('author', {
      default: configOptions.author,
      description: 'Show commits by specific author',
      type: 'string',
    })
    .option('maxNumber', {
      default: configOptions.maxNumber,
      description: 'Number of commits to choose from',
      alias: ['number', 'n'],
      type: 'number',
    })
    .option('dryRun', {
      default: false,
      description: 'Perform backport without pushing to Github',
      type: 'boolean',
    })
    .option('editor', {
      default: configOptions.editor,
      description: 'Editor to be opened during conflict resolution',
      type: 'string',
    })
    .option('fork', {
      default: configOptions.fork,
      description: 'Create backports in fork or origin repo',
      type: 'boolean',
    })
    .option('gitHostname', {
      hidden: true,
      default: configOptions.gitHostname,
      description: 'Hostname for Github',
      type: 'string',
    })
    .option('githubApiBaseUrlV3', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV3,
      description: `Base url for Github's REST (v3) API`,
      type: 'string',
    })
    .option('githubApiBaseUrlV4', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV4,
      description: `Base url for Github's GraphQL (v4) API`,
      type: 'string',
    })
    .option('mainline', {
      description:
        'Parent id of merge commit. Defaults to 1 when supplied without arguments',
      type: 'number',
      coerce: (mainline) => {
        // `--mainline` (default to 1 when no parent is given)
        if (mainline === undefined) {
          return 1;
        }

        // use specified mainline parent
        if (Number.isInteger(mainline)) {
          return mainline as number;
        }

        // Invalid value provided
        throw new Error(`--mainline must be an integer. Received: ${mainline}`);
      },
    })
    .option('multiple', {
      default: configOptions.multiple,
      description: 'Select multiple branches/commits',
      type: 'boolean',
    })
    .option('multipleCommits', {
      default: configOptions.multipleCommits,
      description: 'Backport multiple commits',
      type: 'boolean',
    })
    .option('multipleBranches', {
      default: configOptions.multipleBranches,
      description: 'Backport to multiple branches',
      type: 'boolean',
    })
    .option('noVerify', {
      default: configOptions.noVerify,
      description: 'Bypasses the pre-commit and commit-msg hooks',
      type: 'boolean',
    })
    .option('path', {
      default: configOptions.path,
      description: 'Only list commits touching files under the specified path',
      alias: 'p',
      type: 'string',
    })
    .option('prTitle', {
      default: configOptions.prTitle,
      description: 'Title of pull request',
      alias: 'title',
      type: 'string',
    })
    .option('prDescription', {
      default: configOptions.prDescription,
      description: 'Description to be added to pull request',
      alias: 'description',
      type: 'string',
    })
    .option('pullNumber', {
      conflicts: ['sha', 'sourcePRsFilter'],
      description: 'Pull request to backport',
      type: 'number',
      alias: 'pr',
    })
    .option('resetAuthor', {
      default: false,
      description: 'Set yourself as commit author',
      type: 'boolean',
    })
    .option('sha', {
      conflicts: ['pullNumber', 'sourcePRsFilter'],
      description: 'Commit sha to backport',
      type: 'string',
      alias: 'commit',
    })
    .option('sourcePRLabels', {
      default: configOptions.sourcePRLabels,
      description: 'Add labels to the source (original) PR',
      type: 'array',
      alias: 'sourcePRLabel',
    })
    .option('sourcePRsFilter', {
      conflicts: ['pullNumber', 'sha'],
      // default: configOptions.githubApiBaseUrlV4,
      description: `Filter source pull requests by a query`,
      alias: 'pr-filter',
      type: 'string',
    })
    .option('sourceBranch', {
      default: configOptions.sourceBranch,
      description: `List commits to backport from another branch than master`,
      type: 'string',
    })
    .option('verify', {
      description: `Opposite of no-verify`,
      type: 'boolean',
    })
    .option('targetBranches', {
      default: [] as string[],
      description: 'Branch(es) to backport to',
      type: 'array',
      alias: ['targetBranch', 'branch', 'b'],
      string: true, // ensure `6.0` is not coerced to `6`
    })
    .option('targetPRLabels', {
      default: configOptions.targetPRLabels,
      description: 'Add labels to the target (backport) PR',
      alias: ['labels', 'label', 'l'],
      type: 'array',
    })
    .option('upstream', {
      default: configOptions.upstream,
      description: 'Name of repository',
      alias: 'up',
      type: 'string',
    })
    .option('username', {
      default: configOptions.username,
      description: 'Github username',
      type: 'string',
    })
    .option('verbose', {
      default: false,
      description: 'Show additional debug information',
      type: 'boolean',
    })
    .alias('version', 'v')
    .alias('version', 'V')
    .help()
    .epilogue(
      'For bugs, feature requests or questions: https://github.com/sqren/backport/issues\nOr contact me directly: https://twitter.com/sorenlouv'
    ).argv;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  const { $0, _, verify, ...rest } = cliArgs;

  return {
    ...rest,
    accessToken: cliArgs.accessToken || configOptions.accessToken, // accessToken should not be displayed in yargs help menu
    branchLabelMapping: configOptions.branchLabelMapping, // not available as cli argument
    multipleBranches: cliArgs.multipleBranches || cliArgs.multiple,
    multipleCommits: cliArgs.multipleCommits || cliArgs.multiple,
    noVerify: verify ?? rest.noVerify, // `verify` is a cli-only flag to flip the default of `no-verify`
    targetBranchChoices: configOptions.targetBranchChoices, // not available as cli argument
  };
}
