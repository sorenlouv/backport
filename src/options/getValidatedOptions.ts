import isEmpty from 'lodash.isempty';
import { HandledError } from '../services/HandledError';
import { getGlobalConfigPath } from '../services/env';
import { OptionsFromCliArgs } from './cliArgs';

const GLOBAL_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson';

const PROJECT_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson';

export type ValidatedOptions = ReturnType<typeof getValidatedOptions>;
export function getValidatedOptions({
  // omit `upstream` from options
  upstream,
  ...options
}: OptionsFromCliArgs) {
  // ensure accessToken and username are given
  if (!options.accessToken || !options.username) {
    const globalConfigPath = getGlobalConfigPath();
    throw new HandledError(
      `Please update your config file: ${globalConfigPath}.\nIt must contain a valid "username" and "accessToken".\n\nRead more: ${GLOBAL_CONFIG_DOCS_LINK}`
    );
  }

  // ensure targetBranches or targetBranchChoices are given
  if (isEmpty(options.targetBranches) && isEmpty(options.targetBranchChoices)) {
    throw new HandledError(
      `You must specify a target branch\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--branch 6.1"`
    );
  }

  // ensure a valid upstream is given
  const [repoOwner, repoName] = (upstream ?? '').split('/');
  if (!repoOwner || !repoName) {
    throw new HandledError(
      `You must specify a valid Github repository\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--upstream elastic/kibana"`
    );
  }

  return {
    ...options,

    // no longer optional
    accessToken: options.accessToken,
    username: options.username,

    // split upstream
    repoName,
    repoOwner,

    // define author
    author: options.author || options.username,
    all: options.author ? false : options.all,
  };
}
