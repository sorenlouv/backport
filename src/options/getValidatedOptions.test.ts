import { OptionsFromCliArgs } from './cliArgs';
import { getValidatedOptions } from './getValidatedOptions';

describe('getValidatedOptions', () => {
  const validOptions: OptionsFromCliArgs = {
    accessToken: 'myAccessToken',
    all: false,
    assignees: [],
    author: undefined,
    autoFixConflicts: undefined,
    branchLabelMapping: undefined,
    dryRun: false,
    editor: 'code',
    fork: true,
    gitHostname: 'github.com',
    githubApiBaseUrlV3: 'https://api.github.com',
    githubApiBaseUrlV4: 'https://api.github.com/graphql',
    mainline: undefined,
    maxNumber: 10,
    multipleBranches: true,
    multipleCommits: false,
    noVerify: true,
    path: undefined,
    prDescription: undefined,
    prTitle: 'myPrTitle',
    pullNumber: undefined,
    resetAuthor: false,
    sha: undefined,
    sourceBranch: 'mySourceBranch',
    sourcePRLabels: [],
    prFilter: undefined,
    targetBranchChoices: [],
    targetBranches: ['branchA'],
    targetPRLabels: [],
    upstream: 'elastic/kibana',
    username: 'sqren',
    verbose: false,
  };

  describe('should not throw', () => {
    it('when all options are valid and `branches` is given', () => {
      expect(() => getValidatedOptions(validOptions)).not.toThrow();
    });

    it('when `targetBranchChoices` is given instead of `targetBranches`', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          targetBranchChoices: [{ name: 'branchA' }],
          targetBranches: [],
        })
      ).not.toThrow();
    });
  });

  describe('should throw', () => {
    it('when accessToken is missing', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          accessToken: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when accessToken is empty', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          accessToken: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when both `targetBranches` and `targetBranchChoices` are missing', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          targetBranchChoices: [],
          targetBranches: [],
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is missing', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          upstream: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is empty', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          upstream: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is missing', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          username: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is empty', () => {
      expect(() =>
        getValidatedOptions({
          ...validOptions,
          username: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
