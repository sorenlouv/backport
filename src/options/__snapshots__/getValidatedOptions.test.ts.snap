// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getValidatedOptions should throw when accessToken is empty 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`getValidatedOptions should throw when accessToken is missing 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`getValidatedOptions should throw when both \`targetBranches\` and \`targetBranchChoices\` are missing 1`] = `
"You must specify a target branch

You can specify it via either:
 - Config file (recommended): \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--branch 6.1\\""
`;

exports[`getValidatedOptions should throw when upstream is empty 1`] = `
"You must specify a valid Github repository

You can specify it via either:
 - Config file (recommended): \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`getValidatedOptions should throw when upstream is missing 1`] = `
"You must specify a valid Github repository

You can specify it via either:
 - Config file (recommended): \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`getValidatedOptions should throw when username is empty 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`getValidatedOptions should throw when username is missing 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson"
`;
